File: python-reference/src/ft_cli/commands/friends.py
from __future__ import annotations

from datetime import date

import typer

from ft_cli import auth, config
from ft_cli.client import ApiError, Client
from ft_cli.commands.active import fetch_active_locations, resolve_campus
from ft_cli.config import Friend, load_friends, save_friends
from ft_cli.render import (
    console,
    err_console,
    friends_online_table,
    friends_table,
)

app = typer.Typer(help="Local friends watchlist (stored on this machine only).")


@app.command()
def add(login: str = typer.Argument(..., help="42 login to add")) -> None:
    """Validate the login against the API, then append it to the local list."""
    login = login.strip().lower()
    friends = load_friends()
    if friends.has(login):
        console.print(f"[dim]{login} is already in your friends list[/dim]")
        return

    try:
        with Client() as c:
            user = c.get_json(f"/v2/users/{login}")
    except ApiError as e:
        if e.status == 404:
            err_console.print(f"login not found on intra: {login}")
            raise typer.Exit(code=1)
        err_console.print(str(e))
        raise typer.Exit(code=1)
    except auth.AuthError as e:
        err_console.print(str(e))
        raise typer.Exit(code=1)

    friends.add(
        Friend(
            login=login,
            display_name=user.get("displayname") or user.get("usual_full_name") or "",
            added=date.today().isoformat(),
        )
    )
    save_friends(friends)
    console.print(f":heavy_plus_sign: added [cyan]{login}[/cyan]")


@app.command()
def remove(login: str = typer.Argument(..., help="42 login to remove")) -> None:
    """Remove a login from your local friends list."""
    login = login.strip().lower()
    friends = load_friends()
    if friends.remove(login):
        save_friends(friends)
        console.print(f":heavy_minus_sign: removed [cyan]{login}[/cyan]")
    else:
        err_console.print(f"{login} is not in your friends list")
        raise typer.Exit(code=1)


@app.command("list")
def list_cmd() -> None:
    """Show all friends in your local list."""
    friends = load_friends()
    friends_table(
        [
            {"login": f.login, "display_name": f.display_name, "added": f.added}
            for f in friends.friends
        ]
    )


@app.command()
def online(
    campus: str | None = typer.Option(
        None, "--campus", "-c", help="campus slug (defaults to your saved campus)"
    ),
) -> None:
    """Show which of your friends are currently active at the campus."""
    friends = load_friends()
    if not friends.friends:
        console.print("[dim]no friends yet — add some with `42 friends add <login>`[/dim]")
        return

    friend_logins = {f.login for f in friends.friends}
    try:
        with Client() as c:
            _slug, campus_id = resolve_campus(c, campus)
            locations = fetch_active_locations(c, campus_id, limit=1000)
    except (auth.AuthError, ApiError) as e:
        err_console.print(str(e))
        raise typer.Exit(code=1)

    matched = [
        loc for loc in locations if (loc.get("user") or {}).get("login") in friend_logins
    ]
    active_logins = {(loc.get("user") or {}).get("login") for loc in matched}
    offline = list(friend_logins - active_logins)
    friends_online_table(matched, offline)

File: src/commands/friends.ts
import kleur from "kleur";

import { AuthError } from "../auth.js";
import { apiGet, ApiError } from "../client.js";
import {
  friendsAdd,
  friendsHas,
  friendsRemove,
  loadFriends,
  saveFriends,
} from "../config.js";
import { err, friendsOnlineTable, friendsTable } from "../render.js";
import { fetchActiveLocations, resolveCampus } from "./active.js";

function today(): string {
  return new Date().toISOString().slice(0, 10);
}

export async function friendsAddCmd(login: string): Promise<void> {
  login = login.trim().toLowerCase();
  const friends = loadFriends();
  if (friendsHas(friends, login)) {
    console.log(kleur.dim(`${login} is already in your friends list`));
    return;
  }
  try {
    const user = await apiGet<any>(`/v2/users/${encodeURIComponent(login)}`);
    friendsAdd(friends, {
      login,
      display_name: user.displayname ?? user.usual_full_name ?? "",
      added: today(),
    });
    saveFriends(friends);
    console.log(kleur.green("+") + " added " + kleur.cyan(login));
  } catch (e) {
    if (e instanceof ApiError && e.status === 404) {
      err(`login not found on intra: ${login}`);
      process.exit(1);
    }
    if (e instanceof ApiError || e instanceof AuthError) {
      err(e.message);
      process.exit(1);
    }
    throw e;
  }
}

export function friendsRemoveCmd(login: string): void {
  login = login.trim().toLowerCase();
  const friends = loadFriends();
  if (friendsRemove(friends, login)) {
    saveFriends(friends);
    console.log(kleur.red("-") + " removed " + kleur.cyan(login));
  } else {
    err(`${login} is not in your friends list`);
    process.exit(1);
  }
}

export function friendsListCmd(): void {
  const friends = loadFriends();
  friendsTable(friends.friends);
}

export async function friendsOnlineCmd(opts: { campus?: string }): Promise<void> {
  const friends = loadFriends();
  if (friends.friends.length === 0) {
    console.log(
      kleur.dim("no friends yet — add some with `japonette friends add <login>`"),
    );
    return;
  }
  const wanted = new Set(friends.friends.map((f) => f.login));
  try {
    const { id } = await resolveCampus(opts.campus);
    const locs = await fetchActiveLocations(id, 1000);
    const matched = locs.filter((l) => wanted.has(l?.user?.login));
    const activeLogins = new Set(matched.map((l) => l?.user?.login));
    const offline = [...wanted].filter((l) => !activeLogins.has(l));
    friendsOnlineTable(matched, offline);
  } catch (e) {
    if (e instanceof ApiError || e instanceof AuthError) {
      err(e.message);
      process.exit(1);
    }
    if (e instanceof Error) {
      err(e.message);
      process.exit(1);
    }
    throw e;
  }
}
